File: src/ships/constructionBilgeKeels.ts
import type { ConstructionBilgeKeels, Vec3 } from './blueprint';

export const defaultBilgeKeels = (beam: number): ConstructionBilgeKeels => ({
  version: 1, enabled: true, start: .25, end: .75,
  widthM: Math.round(Math.max(.1, Math.min(2, beam * .035)) * 100) / 100,
  thicknessM: .02, placement: .35,
});
export function bilgeKeelError(k: ConstructionBilgeKeels): string | undefined {
  if (k.version !== 1 || typeof k.enabled !== 'boolean') return 'Unsupported bilge keel settings.';
  if (![k.start, k.end, k.widthM, k.thicknessM, k.placement].every(Number.isFinite)) return 'Enter finite bilge keel dimensions.';
  if (k.start < .02 - 1e-10 || k.end > .98 + 1e-10 || k.end - k.start < .02 - 1e-10) return 'Keep bilge keels between 2% and 98% of hull length, spanning at least 2%.';
  if (k.widthM < .05 || k.widthM > 3 || k.thicknessM < .005 || k.thicknessM > .1 || k.thicknessM > k.widthM) return 'Use bilge keel width 0.05–3 m and thickness 5–100 mm, no thicker than the width.';
  if (k.placement < .05 || k.placement > .8) return 'Place bilge keels 5–80% around the outline from keel to deck.';
}

/** Display counterpart of the native recipe. Roots follow the actual triangulated
 * hull panels, including their diagonals; a small embedded edge seats the plates. */
export function bilgeKeelFaces(rings: Vec3[][], stations: number[], contours: number[], k?: ConstructionBilgeKeels, cut = 1): Vec3[][] {
  if (!k?.enabled || bilgeKeelError(k) || cut <= k.start) return [];
  const q = 4 + k.placement * 4;
  let edge = (contours.length - 1) / 2;
  while (edge < contours.length - 2 && contours[edge + 1] < q) edge++;
  const u = (q - contours[edge]) / (contours[edge + 1] - contours[edge]);
  const end = Math.min(k.end, cut), taper = (k.end - k.start) * .12;
  const times = [k.start, end, k.start + taper, k.end - taper, ...stations];
  for (let i = 0; i < stations.length - 1; i++) times.push(stations[i] + u * (stations[i + 1] - stations[i]));
  const ts = [...new Set(times.filter(t => t >= k.start && t <= end))].sort((a,b) => a-b);
  const mix = (a: Vec3, b: Vec3, t: number) => a.map((v,i) => v + (b[i]-v)*t) as Vec3;
  const sections = ts.map(t => {
    let j = 0;
    while (j < stations.length - 2 && stations[j + 1] < t) j++;
    const f = (t - stations[j]) / (stations[j + 1] - stations[j]);
    const a = rings[j][edge], b = rings[j][edge+1], c = rings[j+1][edge], d = rings[j+1][edge+1];
    const root = a.map((v,i) => u >= f ? v*(1-u)+b[i]*(u-f)+d[i]*f : v*(1-f)+d[i]*u+c[i]*(f-u)) as Vec3;
    const low = mix(a,c,f), high = mix(b,d,f), dx = high[0]-low[0], dy = high[1]-low[1], length = Math.hypot(dx,dy) || 1;
    const nx = dy/length, ny = -dx/length;
    const width = k.widthM * Math.max(.08, Math.min(1, (t-k.start)/taper, (k.end-t)/taper));
    return [[-k.thicknessM,-.5], [width,-.5], [width,.5], [-k.thicknessM,.5]].map(([w,s]) =>
      [root[0]+nx*w-ny*k.thicknessM*s, root[1]+ny*w+nx*k.thicknessM*s, root[2]] as Vec3);
  });
  const faces: Vec3[][] = [];
  const triangle = (a: Vec3,b: Vec3,c: Vec3) => {
    faces.push([a,b,c], [c,b,a].map(p => [-p[0],p[1],p[2]]));
  };
  for (let j=0;j<sections.length-1;j++) for (let i=0;i<4;i++) {
    const a=sections[j][i],b=sections[j][(i+1)%4],c=sections[j+1][i],d=sections[j+1][(i+1)%4];
    triangle(a,b,c); triangle(b,d,c);
  }
  const first=sections[0],last=sections.at(-1)!;
  triangle(first[2],first[1],first[0]); triangle(first[3],first[2],first[0]);
  triangle(last[0],last[1],last[2]); triangle(last[0],last[2],last[3]);
  return faces;
}

File: crates/naval-sim/src/construction_bilge_keels.rs
//! Visual plates, deliberately separate from buoyant cells, armor and loading.
use crate::{construction_geometry as cg, definition::*, geometry::*};

pub fn validate(k: &ConstructionBilgeKeels) -> Result<(), String> {
    if k.version != 1. {
        return Err("Unsupported bilge keel settings".into());
    }
    if ![k.start, k.end, k.width_m, k.thickness_m, k.placement]
        .iter()
        .all(|n| n.is_finite())
    {
        return Err("Enter finite bilge keel dimensions".into());
    }
    if k.start < 0.02 - 1e-10 || k.end > 0.98 + 1e-10 || k.end - k.start < 0.02 - 1e-10 {
        return Err(
            "Keep bilge keels between 2% and 98% of hull length, spanning at least 2%".into(),
        );
    }
    if !(0.05..=3.).contains(&k.width_m)
        || !(0.005..=0.1).contains(&k.thickness_m)
        || k.thickness_m > k.width_m
    {
        return Err(
            "Use bilge keel width 0.05–3 m and thickness 5–100 mm, no thicker than the width"
                .into(),
        );
    }
    if !(0.05..=0.8).contains(&k.placement) {
        return Err("Place bilge keels 5–80% around the outline from keel to deck".into());
    }
    Ok(())
}

pub fn surfaces(p: &ConstructionPrimitive, paint: &str) -> Vec<ConstructionSurface> {
    let Some(h) = &p.custom_hull else {
        return vec![];
    };
    let Some(k) = &h.bilge_keels else {
        return vec![];
    };
    if !k.enabled || validate(k).is_err() {
        return vec![];
    }
    let mut local = p.clone();
    local.position = [0.; 3];
    local.rotation_deg = 0.;
    local.tilt = None;
    let n = h.stations[0].points.len();
    let contour = |i: usize| {
        h.stations[0].points[i]
            .contour
            .unwrap_or(i as f64 * 8. / (n - 1) as f64)
    };
    let q = 4. + k.placement * 4.;
    let mut edge = (n - 1) / 2;
    while edge < n - 2 && contour(edge + 1) < q {
        edge += 1;
    }
    let u = (q - contour(edge)) / (contour(edge + 1) - contour(edge));
    let rings: Vec<Vec<Vec3>> = h
        .stations
        .iter()
        .map(|s| {
            s.points
                .iter()
                .enumerate()
                .map(|(i, v)| {
                    crate::construction_custom_hull::transform(&local, s.t, v, contour(i))
                })
                .collect()
        })
        .collect();
    let taper = (k.end - k.start) * 0.12;
    let mut times = vec![k.start, k.end, k.start + taper, k.end - taper];
    times.extend(h.stations.iter().map(|s| s.t));
    times.extend(
        h.stations
            .windows(2)
            .map(|s| s[0].t + u * (s[1].t - s[0].t)),
    );
    times.retain(|t| *t >= k.start && *t <= k.end);
    times.sort_by(f64::total_cmp);
    times.dedup();
    let sections: Vec<Vec<Vec3>> = times
        .iter()
        .map(|&t| {
            let mut j = 0;
            while j < h.stations.len() - 2 && h.stations[j + 1].t < t {
                j += 1;
            }
            let f = (t - h.stations[j].t) / (h.stations[j + 1].t - h.stations[j].t);
            let [a, b, c, d] = [
                rings[j][edge],
                rings[j][edge + 1],
                rings[j + 1][edge],
                rings[j + 1][edge + 1],
            ];
            let root = if u >= f {
                add(add(scale(a, 1. - u), scale(b, u - f)), scale(d, f))
            } else {
                add(add(scale(a, 1. - f), scale(d, u)), scale(c, f - u))
            };
            let low = add(a, scale(sub(c, a), f));
            let high = add(b, scale(sub(d, b), f));
            let dx = high[0] - low[0];
            let dy = high[1] - low[1];
            let length = dx.hypot(dy);
            let length = if length > 0. { length } else { 1. };
            let nx = dy / length;
            let ny = -dx / length;
            let width = k.width_m
                * ((t - k.start) / taper)
                    .min((k.end - t) / taper)
                    .clamp(0.08, 1.);
            [
                (-k.thickness_m, -0.5),
                (width, -0.5),
                (width, 0.5),
                (-k.thickness_m, 0.5),
            ]
            .iter()
            .map(|&(w, s)| {
                [
                    root[0] + nx * w - ny * k.thickness_m * s,
                    root[1] + ny * w + nx * k.thickness_m * s,
                    root[2],
                ]
            })
            .collect()
        })
        .collect();
    let mut faces = vec![];
    let mut triangle = |a: Vec3, b: Vec3, c: Vec3| {
        faces.push(vec![a, b, c]);
        faces.push([c, b, a].iter().map(|v| [-v[0], v[1], v[2]]).collect());
    };
    for pair in sections.windows(2) {
        for i in 0..4 {
            let [a, b, c, d] = [
                pair[0][i],
                pair[0][(i + 1) % 4],
                pair[1][i],
                pair[1][(i + 1) % 4],
            ];
            triangle(a, b, c);
            triangle(b, d, c);
        }
    }
    let first = &sections[0];
    let last = sections.last().unwrap();
    triangle(first[2], first[1], first[0]);
    triangle(first[3], first[2], first[0]);
    triangle(last[0], last[1], last[2]);
    triangle(last[0], last[2], last[3]);
    faces
        .into_iter()
        .enumerate()
        .map(|(i, face)| {
            let vertices: Vec<_> = face
                .into_iter()
                .map(|v| crate::construction_orientation::point(p, v))
                .collect();
            ConstructionSurface {
                id: format!("{}:bilge-keel:{i}", p.id),
                primitive_id: p.id.clone(),
                face: "bilge-keel".into(),
                panel_id: Some(format!("bilge-keel:{i}")),
                normal: cg::normal(&vertices),
                area_m2: cg::area(&vertices),
                vertices,
                thickness_mm: k.thickness_m * 1000.,
                material: "steel".into(),
                paint: paint.into(),
                open: false,
            }
        })
        .collect()
}
